app/src/rx_fastmetadata_app.c

```diff
@@ -4,12 +4,152 @@
+      if (ret) {
+        session->errors_count++;
+        err("%s() FAIL: reference file comparison (2).\n", __func__);
+      } else {
+        dbg("%s() PASS: reference file comparison (2).\n", __func__);
+
+        /* Verify last 0-3 bytes of frame (filled with zero's) */
+        ret = memcmp(&(((uint8_t*)frame)[st41_ref_remaining_length]), (void*)&last_zeros,
+                     frame_size - st41_ref_remaining_length);
+        if (ret) {
+          session->errors_count++;
+          err("%s() FAIL: reference file comparison (3).\n", __func__);
+        } else {
+          dbg("%s() PASS: reference file comparison (3).\n", __func__);
+        }
+      }
+    }
+  }
+
+#ifdef DEBUG /* print out of frame and reference file if error */
+  if (ret) {
+    err("%s() FRAME START>>", __func__);
+    for (int i = 0; i < frame_size; i++) {
+      err("%c", ((char*)frame)[i]);
+    }
+    err("<<END FRAME.\n\n");
+
+    err("%s() REFERENCE START>>", __func__);
+    for (int i = 0; i < frame_size; i++) {
+      err("%c", ((char*)(session->st41_ref_cursor))[i]);
+    }
+    err("<<END REFERENCE.\n\n");
+  }
+#endif /* DEBUG */
+
+  /* calculate new reference frame */
+  session->st41_ref_cursor += frame_size;
+  if (session->st41_ref_cursor >= session->st41_ref_end)
+    session->st41_ref_cursor = session->st41_ref_begin;
+
+  return ret;
+}
+
+static void app_rx_fmd_handle_rtp(struct st_app_rx_fmd_session* s, void* usrptr) {
+  struct st41_rtp_hdr* hdr = (struct st41_rtp_hdr*)usrptr;
+  uint8_t* payload = (uint8_t*)(&hdr[1]);
+
+  hdr->swaped_st41_hdr_chunk = ntohl(hdr->swaped_st41_hdr_chunk);
+
+  /* Testing data_item_type */
+  if (s->st41_dit != 0xffffffff) {
+    if (hdr->st41_hdr_chunk.data_item_type != s->st41_dit) {
+      s->errors_count++;
+      err("%s(%d) FAIL: hdr->st41_hdr_chunk.data_item_type=%u, expected to be %u.\n",
+          __func__, s->idx, hdr->st41_hdr_chunk.data_item_type, s->st41_dit);
+    } else {
+      dbg("%s(%d) PASS: hdr->st41_hdr_chunk.data_item_type=%u, expected to be %u.\n",
+          __func__, s->idx, hdr->st41_hdr_chunk.data_item_type, s->st41_dit);
+    }
+  }
+
+  /* Testing data_item K-bit */
+  if (s->st41_k_bit != 0xff) {
+    if (hdr->st41_hdr_chunk.data_item_k_bit != s->st41_k_bit) {
+      s->errors_count++;
+      err("%s(%d) FAIL: hdr->st41_hdr_chunk.data_item_k_bit=%u, expected to be %u.\n",
+          __func__, s->idx, hdr->st41_hdr_chunk.data_item_k_bit, s->st41_k_bit);
+    } else {
+      dbg("%s(%d) PASS: hdr->st41_hdr_chunk.data_item_k_bit=%u, expected to be %u.\n",
+          __func__, s->idx, hdr->st41_hdr_chunk.data_item_k_bit, s->st41_k_bit);
+    }
+  }
 
   s->stat_frame_total_received++;
   if (!s->stat_frame_first_rx_time)
     s->stat_frame_first_rx_time = st_app_get_monotonic_time();
+
+  /* Compare each packet with reference (part by part) */
+  if (s->st41_ref_fd > 0)
+    app_rx_fmd_compare_with_ref(s, payload, hdr->st41_hdr_chunk.data_item_length * 4);
```

Comment:
woudn't data_item_length_bytes work here a lil bit better ?

Reply:
It would. :) But then I would have to create this particular variable just for that usage. And nowhere else.

Reply:
oooh i thought it was part of this structure i understand,
mby a comment ? 
Why we are multiplying via 4 ? 
It would be more clear to have a reason there ( or change the name of the variable to say that)

Reply:
So should I add this variable?

Reply:
I will add the comment.